xarray/core/indexing.py

```diff
@@ -166,6 +184,17 @@ def convert_label_indexer(index, label, index_name='', method=None,
             raise KeyError('cannot represent labeled-based slice indexer for '
                            'dimension %r with a slice over integer positions; '
                            'the index is unsorted or non-unique')
+
+    elif is_dict_like(label):
+        if not isinstance(index, pd.MultiIndex):
+            raise ValueError('cannot use a dict-like object for selection on a '
+                             'dimension that does not have a MultiIndex')
+        indexer, new_index = index.get_loc_level(tuple(label.values()),
+                                                 level=tuple(label.keys()))
+
+    elif _is_nested_tuple(label, index):
+        indexer = index.get_locs(label)
+
```

Comment:
Doing some tests, it seems like `get_loc_level` properly drops levels from a MultiIndex, but doesn't handle dropping the level entirely. For that, we need to use `get_loc`.

Good:
```
In [11]: idx = pd.MultiIndex.from_product([['a', 'b'], [1, 2], [-1, -2]])

In [12]: idx.get_loc_level(('a', 1), [0, 1])
Out[12]:
(array([ True,  True, False, False, False, False, False, False], dtype=bool),
 Int64Index([-1, -2], dtype='int64'))

In [13]: idx.get_loc_level(('a', -1), [0, 2])
Out[13]:
(array([ True, False,  True, False, False, False, False, False], dtype=bool),
 Int64Index([1, 2], dtype='int64'))

In [15]: idx.get_loc_level(('a',), [0])
Out[15]:
(array([ True,  True,  True,  True, False, False, False, False], dtype=bool),
 MultiIndex(levels=[[1, 2], [-2, -1]],
            labels=[[0, 0, 1, 1], [1, 0, 1, 0]]))

In [16]: idx.get_loc_level((1,), [1])
Out[16]:
(array([ True,  True, False, False,  True,  True, False, False], dtype=bool),
 MultiIndex(levels=[['a', 'b'], [-2, -1]],
            labels=[[0, 0, 1, 1], [1, 0, 1, 0]]))
```

Bad:
```
In [14]: idx.get_loc_level(('a', 1, -1), [0, 1, 2])
Out[14]:
(array([ True, False, False, False, False, False, False, False], dtype=bool),
 MultiIndex(levels=[['a', 'b'], [1, 2], [-2, -1]],
            labels=[[0], [0], [1]]))
```

Good:
```
In [17]: idx.get_loc(('a', 1, -1))
Out[17]: 0
```